docs/design/authentication_policy.md

```diff
@@ -0,0 +1,40 @@
+# Envoy Gateway Authentication Policy API
+
+## Overview
+
+This authen policy is to declare the authentication mechanisms, to be enforce on connection and request going though Envoy Gateway. This includes the credential (X.509, JWT, etc), parameters (cipher suites, key algorithms)
+The policy is similar to [OpenAPI 3.1 security objects](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#securitySchemeObject) without the API key part, and should be easily translatable from it with some additions.
+
+## Authentication mechanisms
+This policy should support the following authentication mechanisms:
+- JWT Bearer Token
+- mutualTLS (client certificate)
+- OAuth2
+- OIDC
+- External authentication
+
+In the first phase, Envoy Gateway will implement JWT Bearer Token authentication.
+
+In general those policy translates into Envoy HTTP filters at HTTP connection manager level, and route specific settings will be applied for each route. These APIs are expressed in a Policy CRD and attached to Gateway API resources with [Policy Attachement](https://gateway-api.sigs.k8s.io/references/policy-attachment/).
```

Comment:
re: `route specific settings will be applied for each route`

The Gateway API Policy `targetRef` field does not yet have a way to specify a sub-section of a resource, so applying a Policy to an individual route *rule* is not "directly" possible, unless you require users to separate that route rule into its own HTTPRoute resource (in this case you might have a Policy target a Gateway and another target an attached HTTPRoute that specifies some overrides just for it)

Am I understanding the intent here properly? Might be good to flesh out some more examples on how this might work

Reply:
see also: https://github.com/kubernetes-sigs/gateway-api/discussions/1489

Reply:
thanks for writing down that in depth analysis ! 
and surfacing some limitations with the PolicyAttachment that exist today.
Hoping those get incorporated soon, @youngnick has already acknowledged this !

Until then, since we plan on adding new features such as `authentication` in our next v0.3.0 release, imho we should peruse the existing spec for now

Reply:
From the Gateway API [Policy Attachment spec](https://gateway-api.sigs.k8s.io/references/policy-attachment/):

> Although Policy attachment can be used to target an entire Route or Backend, it cannot currently be used to target specific Route rules or backend references.

Based on this information, it appears that an authentication API fits better as an [implementation-specific filter](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.HTTPRouteFilter) since HTTPRoute rules may require different authentication specs. In the future, EG can add GWAPI Policy Attachment support for authentication defaulting and overrides.

Reply:
HTTPRoute Filters themselves also have some drawbacks as they may not directly translate to Envoy HTTP Filters, given the scope they are configurable at.

HTTPRoute Filters can be configured per-Route Rule or per-Backend and Envoy HTTP Filters mostly have the bulk of their configuration as part of the Filter in a per-virtualhost HTTPConnectionManager FilterChain. I say "mostly" here because many filters enable per-Route enablement/disablement, but most of the configuration (e.g. JWT provider specification: https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/jwt_authn/v3/config.proto#envoy-v3-api-msg-extensions-filters-http-jwt-authn-v3-jwtauthentication) lives at the higher level Filter object.

We may be able to have per-route JWT providers we collect from HTTPRoute Filter objects and then configure on the Envoy JWT HTTP Filter, but I haven't actually tried it so not sure how complex that will get.

Contour implemented this recently in HTTPProxy here:
- https://github.com/projectcontour/contour/blob/4badccc7267aeec2644f9122ed7ca1547934a5b5/design/jwt-verification-design.md
- https://projectcontour.io/docs/v1.23.0/config/jwt-verification/

Curious to see your thoughts @lizan if you've gotten a chance to spike this out as a Policy vs. a Filter

Reply:
> HTTPRoute Filters can be configured per-Route Rule or per-Backend

#733 implements authN as an [HTTPRoute filter](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.HTTPRouteFilter).

> We may be able to have per-route JWT providers we collect from HTTPRoute Filter objects and then configure on the Envoy JWT HTTP Filter, but I haven't actually tried it so not sure how complex that will get.

I am going to test this as an approach for implementing authN. @lizan please comment if you have any suggestions.